text/trace/0189-instrumentation-layers-suppression.md

```diff
@@ -0,0 +1,222 @@
+# Instrumentation Layers and Suppression
+
+This document describes approach for tracing instrumentation layers and suppressing duplicate layers.
+
+## Motivation
+
+- Provide clarity for instrumentation layers: e.g. DB calls on top of REST API
+- Give mechanism to suppress instrumentation layers for the same convention: e.g. multiple instrumented HTTP clients using each other.

```

Comment:
Why are there cases of multiple instrumentation HTTP clients using each other? That sounds like an odd configuration of clients. I wonder if when combining these clients, one can disable instrumentation of the other?

The DB call on top fo a REST call -- you propose to suppress the REST instrumentation, I take it? (Could the DB be configured with an un-instrumented REST client, instead?) 

I wonder if it would be possible to replace `InstrumentationType` with `InstrumentationLibrary.Name`. Could you accomplish the same functional behavior, if for all spans with a non-remote parent you knew the InstrumentationLibrary of the nearest not-suppressed ancestor? It might make the proposal both simpler and more general: InstrumentationLibrary is already defined; it means you don't need an expandable enum.

Reply:
> Why are there cases of multiple instrumentation HTTP clients using each other? 

1. One case is manual/native + auto: https://github.com/open-telemetry/opentelemetry-specification/issues/1767
2. Also, HTTP clients test to use each other (python, js, java all have one or another way to suppress HTTP instrumentation)
- in some cases, they are pluggable, i.e. higher level don't really know if the lower level is instrumented - this is more popular in SERVER spans
- in some cases,  I guess higher-level clients instrumentation is not necessary

> The DB call on top fo a REST call -- you propose to suppress the REST instrumentation, I take it? (Could the DB be configured with an un-instrumented REST client, instead?)

I suggest that 
1. DB and rest are two layers, i.e. they are both spans  (and not DB attributes stamped on HTTP span)
2. Depending on the configuration we can either
    - suppress nested HTTP (to reduce noise if users/backends don't want it)
    - not suppress HTTP and have full picture


> InstrumentationType with InstrumentationLibrary.Name

I'd love that, however `InstrumentationLibrary.Name` could be `GoogleHttpClient` or `ApacheHttpClient` while they both implement HTTP semantic conventions. The alternative could be to set `InstrumentationType` on the tracer resource in addition to `InstrumentationLibrary.Name` and version. I'd also put the semantic convention version it follows there.

Reply:
> InstrumentationType with InstrumentationLibrary.Name

Experimented with this idea: added another 'type' parameter to library instrumentation info when tracer is being built:

- it makes tracers type-aware, i.e. tracers should not emit spans of different types. Which makes sense, but a new thing. 
-  it's even more subtle than the span creation option.  App and lib devs would have a hard time paying attention and will end up with multi-type tracers (i.e. no suppression).

I updated the PR https://github.com/lmolkova/opentelemetry-java/pull/1.

Will bring it up at the Tue Instrumentation SIG meeting.

I also entertained the idea of having a single tracer per type and returning noop tracers if one with the same type is registered - but it depends on the tracer creation time and is not stable.